src/mpid/ch4/shm/src/topotree.c

```diff
@@ -126,7 +126,7 @@ int MPIDI_SHM_topotree_get_package_level(int topo_depth, int *max_entries_per_le
         }
     }
     /* STEP 3.3. Determine the package level based on first level (top-down) with #nodes >1 */
-    socket_level = topo_depth;
+    socket_level = 0;   /* if all ranks are in the same index at all levels, just use level 0 */
```

Comment:
I'm having a hard time understanding "all ranks are in the same index at all levels" means. What is the index and what does being in an index mean for a process binding?

Reply:
Refers to this diagram - https://www.open-mpi.org/projects/hwloc/doc/v2.3.0/diagram.png.

When a process' bindset is within, e.g. a single PU, then, its "depth" is at the "PU" depth (level 4 in the diagram but level 6 in current hwloc). If the bindset spans two PUs, then the depth may be e.g. 2 (a single cache locality) or even 0 when none of the smaller locality encapsulates it. Levels and depths are synonyms, except in topo_tree.c, it is reversed in order, which is totally unnecessary.

Within a locality level, each locality is indexed, e.g. core 0, core 1, core 2, .... Index refers to the locality index the process' bindset is within.

Reply:
The commit refers to the case when all processes are within the same locality index at all levels. The old code choose the level at `topo_depth`, which is invalid since it the highest depth `+` 1. It should be just `0`, which is always valid.

Reply:
The code is a mess to read. Refer to https://github.com/pmodels/mpich/pull/5469 see a bit cleaned up version.

Reply:
How does `-bind-to core:2` result in all processes being within the same locality index? Is it just that the machines we run on only have a single topology object above core? Or a limited number of processes?

Reply:
The test only run on 4 processes, all 4 processes are on the same socket. Each process span 2 cores so they don't have a index at core level

Reply:
OK, that makes sense. I'm still not sure if the "same index at all levels" comment helps my understanding or hurts it. 

Ultimately, what I think is happening here is that a single shm region will be created if no suitable split across hardware elements is found within the process binding. Is that right?

Reply:
And the problem was that before, the default case (no suitable split) was to create a shm region too deep in the hierarchy to be useful?

Reply:
Before the failed case have `topo_depth` to `3`, but all 3 levels don't more than 1 splits, so package_level was left as default to `3`. Later when the code trying to access `bindmap[3]` it segfaults, as only index `0, 1, 2` are allocated.

Reply:
OK, I think I get it now. Looks fine.